src/natcap/invest/urban_nature_access.py

```diff
@@ -165,29 +189,289 @@ def execute(args):
+                (population_nodata, 'raw')],
+            'local_op': _greenspace_population_ratio,
+            'target_raster_path': file_registry['greenspace_population_ratio'],
+            'datatype_target': gdal.GDT_Float32,
+            'nodata_target': FLOAT32_NODATA
+        },
+        task_name='2SFCA: Calculate R_j greenspace/population ratio',
+        target_path_list=[file_registry['greenspace_population_ratio']],
+        dependent_task_list=[
+            convolved_population_task,
+            greenspace_reclassification_task,
+        ])
+
+    convolved_greenspace_population_ratio_task = graph.add_task(
+        pygeoprocessing.convolve_2d,
+        kwargs={
+            'signal_path_band': (
+                file_registry['greenspace_population_ratio'], 1),
+            'kernel_path_band': (file_registry['kernel'], 1),
+            'target_path': file_registry['greenspace_supply'],
+            'working_dir': intermediate_dir,
+        },
+        task_name='2SFCA - greenspace supply',
+        target_path_list=[file_registry['greenspace_supply']],
+        dependent_task_list=[
+            kernel_creation_task,
+            greenspace_population_ratio_task,
+        ])
+
     graph.close()
     graph.join()
     LOGGER.info('Finished Urban Nature Access Model')
 
 
+def _greenspace_population_ratio(
+        greenspace_area, convolved_population, greenspace_nodata,
+        population_nodata):
+    """Calculate the greenspace-population ratio R_j.
+
+    Args:
+        greenspace_area (numpy.array): A numpy array representing the area of
+            greenspace in the pixel.  Pixel values will be ``0`` if there is no
+            greenspace.  Pixel values may also match ``greenspace_nodata``.
+        convolved_population (numpy.array): A numpy array where each pixel
+            represents the total number of people within a search radius of
+            each pixel, perhaps adjusted by a search kernel.
+        greenspace_nodata (int or float): The nodata value of the
+            ``greenspace_area`` matrix.
+        population_nodata (int or float): The nodata value of the
+            ``convolved_population`` matrix.
+
+    Returns:
+        A numpy array with the ratio ``R_j`` representing thew
+        greenspace-population ratio with the following constraints:
+
+            * ``convolved_population`` pixels that are numerically close to
+              ``0`` are snapped to ``0`` to avoid unrealistically small
+              denominators in the final ratio.
+            * Any non-greenspace pixels will have a value of ``0.0`` in the
+              output matrix.
+    """
+    # ASSUMPTION: population nodata value is not close to 0.
+    #  Shouldn't be if we're coming from convolution.
+    out_array = numpy.full(
+        greenspace_area.shape, FLOAT32_NODATA, dtype=numpy.float32)
+
+    # Avoid divide-by-zero and very small negative values
+    valid_pixels = (convolved_population > 0)
```

Comment:
Is this possible given `_convolve_and_set_lower_bounds_for_population`?

Reply:
It is possible!  Population can be 0 in a pixel if no one lives there and it happens to be a long ways from a population center (so outside of the convolution kernel or search radius)

But you're also right that we shouldn't be seeing negative values any more, so I've updated that inline comment in e32404c6

Reply:
For whatever reason I didn't think of the 0 case.